libqtile/backend/x11/xcbq.py

```diff
@@ -802,6 +803,52 @@ def grab_pointer(self, owner_events, event_mask, pointer_mode,
     def ungrab_pointer(self):
         self.conn.conn.core.UngrabPointer(xcffib.xproto.Atom._None)
 
+    def paint_borders(self, inner_w, inner_h, borderwidths, bordercolors):
+        if len(bordercolors) == 1 or len(bordercolors) != len(borderwidths):
+            self.set_attribute(borderpixel=bordercolors[0])
+            return
+
+        core = self.conn.conn.core
+        self.borderwidth = sum(borderwidths)
+        outer_w = inner_w + self.borderwidth * 2
+        outer_h = inner_h + self.borderwidth * 2
+
+        pixmap = self.conn.conn.generate_id()
+        core.CreatePixmap(
+            self.conn.default_screen.root_depth, pixmap, self.wid, outer_w, outer_h
+        )
+        gc = self.conn.conn.generate_id()
+        core.CreateGC(gc, pixmap, 0, None)
+
+        coord = 0
+        for i in range(len(borderwidths)):
+            core.ChangeGC(gc, xcffib.xproto.GC.Foreground, [bordercolors[i]])
+            rect = xcffib.xproto.RECTANGLE.synthetic(
+                coord, coord, outer_w - coord * 2, outer_h - coord * 2
+            )
+            core.PolyFillRectangle(pixmap, gc, 1, [rect])
+            coord += borderwidths[i]
+
+        self.set_borderpixmap(pixmap, gc, outer_w, outer_h)
+        core.FreePixmap(pixmap)
+        core.FreeGC(gc)
+
+    def set_borderpixmap(self, pixmap, gc, outer_w, outer_h):
+        core = self.conn.conn.core
+        border = self.conn.conn.generate_id()
+        core.CreatePixmap(
+            self.conn.default_screen.root_depth, border, self.wid, outer_w, outer_h
+        )
+        borderwidth = self.borderwidth
+        most_w = outer_w - borderwidth
+        most_h = outer_h - borderwidth
+        core.CopyArea(pixmap, border, gc, borderwidth, borderwidth, 0, 0, most_w, most_h)
+        core.CopyArea(pixmap, border, gc, 0, 0, most_w, most_h, borderwidth, borderwidth)
+        core.CopyArea(pixmap, border, gc, borderwidth, 0, 0, most_h, most_w, borderwidth)
+        core.CopyArea(pixmap, border, gc, 0, borderwidth, most_w, 0, borderwidth, most_h)
+        core.ChangeWindowAttributes(self.wid, xcffib.xproto.CW.BorderPixmap, [border])
+        core.FreePixmap(border)
```

Comment:
Some similar leakage here. I wonder if we shouldn't add some kind of wrapper for doing this sort of thing to xcffib.

Reply:
try-finally might be the most simple approach - just 4 extra lines, 2 for each of these methods. I took a look at xcffib to see if I could submit a context manager PR but I couldn't wrap my head around how it worked. Maybe we could do the try-finally for the time being, and if an xcffib wrapper seems like a good idea in the future for something else then it could be implemented and then applied here too. What do you think?

Reply:
Happy to do it whatever way you think is best. We can chat about how xcffib works on IRC if you like, but no worries if you think it's simpler to just do here.